numpy/core/src/multiarray/methods.c

```diff
@@ -2694,6 +2696,152 @@ array_complex(PyArrayObject *self, PyObject *NPY_UNUSED(args))
+                             NULL, NULL, NULL))
+    {
+        return NULL;
+    }
+
+    if (stream != Py_None)
+    {
+        PyErr_SetString(PyExc_RuntimeError, "NumPy only supports stream=None.");
+        return NULL;
+    }
+
+    npy_intp itemsize = PyArray_ITEMSIZE(self);
+    int ndim = PyArray_NDIM(self);
+    npy_intp *strides = PyArray_STRIDES(self);
+    npy_intp *shape = PyArray_SHAPE(self);
+
+    for (int i = 0; i < ndim; ++i)
+    {
+        if (strides[i] % itemsize != 0) {
+            PyErr_SetString(PyExc_RuntimeError,
+                            "DLPack only supports strides which are a multiple of itemsize.");
+            return NULL;
+        }
+    }
+
+    DLDataType managed_dtype;
+    PyArray_Descr *dtype = PyArray_DESCR(self);
+    
+    managed_dtype.bits = 8 * itemsize;
+    managed_dtype.lanes = 1;
+    if (PyDataType_ISSIGNED(dtype))
+    {
+        managed_dtype.code = kDLInt;
+    }
+    else if (PyDataType_ISUNSIGNED(dtype))
+    {
+        managed_dtype.code = kDLUInt;
+    }
+    else if (PyDataType_ISFLOAT(dtype))
+    {
+        managed_dtype.code = kDLFloat;
+    }
+    else if (PyDataType_ISCOMPLEX(dtype))
+    {
+        managed_dtype.code = kDLComplex;
+    }
+    else
+    {
+        PyErr_SetString(PyExc_TypeError,
+                        "DLPack only supports signed/unsigned integers, float and complex dtypes.");
+        return NULL;
+    }
+
+    DLManagedTensor *managed = malloc(sizeof(DLManagedTensor));
+    if (managed == NULL)
+    {
+        PyErr_SetString(PyExc_MemoryError,
+                        "Could not allocate the DLManagedTensor struct.");
+        return NULL;
+    }
+
+    managed->dl_tensor.data = PyArray_DATA(self);
+    managed->dl_tensor.device.device_type = kDLCPU;
+    managed->dl_tensor.device.device_id = 0;
+    managed->dl_tensor.dtype = managed_dtype;
+
+
+    int64_t *managed_shape = malloc(sizeof(int64_t) * ndim);
+    if (managed_shape == NULL)
+    {
+        PyErr_SetString(PyExc_MemoryError,
+                        "Could not allocate the DLManagedTensor struct shape.");
+        free(managed);
+        return NULL;
+    }
+
+    int64_t *managed_strides = malloc(sizeof(int64_t) * ndim);
+    if (managed_strides == NULL)
+    {
+        PyErr_SetString(PyExc_MemoryError,
+                        "Could not allocate the DLManagedTensor struct strides.");
+        free(managed);
+        free(managed_shape);
+        return NULL;
+    }
+
+    for (int i = 0; i < ndim; ++i)
+    {
+        managed_shape[i] = shape[i];
+        managed_strides[i] = strides[i];
+    }
+
+    managed->dl_tensor.ndim = ndim;
+    managed->dl_tensor.shape = managed_shape;
+    managed->dl_tensor.strides = managed_strides;
+    managed->dl_tensor.byte_offset = 0;
+    managed->manager_ctx = self;
+    managed->deleter = array_dlpack_deleter;
+
+    PyObject *capsule = PyCapsule_New(managed, NPY_DLPACK_CAPSULE_NAME, array_dlpack_capsule_deleter);
```

Comment:
```suggestion
    PyObject *capsule = PyCapsule_New(managed, NPY_DLPACK_CAPSULE_NAME, NULL);
```

Unsure if the `PyCapsule` needs a deleter.

Reply:
Well, we have to be able to delete it to free the memory, but maybe we need some state as well, to ensure that we don't delete it twice if the consumer already deletes it without deleting the capsule (might be good for non-reference counted Python also?).

I thought I read that there was some capsule renaming involved (to signal that its "invalid" after deletion).  I guess setting the data to NULL or so would be just as well.  (If this is standardized, do that. If not setting data to NULL to guard against deleting twice? Once from python once from the consumer.)